tests/PHPStan/Analyser/nsrt/bug-11311.php

```diff
@@ -29,3 +29,40 @@ function unmatchedAsNullWithOptionalGroup(string $s): void {
 	assertType('array{}|array{string, string|null}', $matches);
 }
 
+function bug11331a(string $url):void {
+	// group a is actually optional as the entire (?:...) around it is optional
+	if (preg_match('{^
+	(?:
+		(?<a>.+)
+	)?
+	(?<b>.+)}mix', $url, $matches, PREG_UNMATCHED_AS_NULL)) {
+		assertType('array{0: string, a: string|null, 1: string|null, b: string, 2: string}', $matches);
+	}
+}
+
+function bug11331b(string $url):void {
+	if (preg_match('{^
+	(?:
+		(?<a>.+)
+	)?
+	(?<b>.+)?}mix', $url, $matches, PREG_UNMATCHED_AS_NULL)) {
+		assertType('array{0: string, a: string|null, 1: string|null, b: string|null, 2: string|null}', $matches);
+	}
+}
+
+function bug11331c(string $url):void {
+	if (preg_match('{^
+	(?:
+		(?:https?|git)://([^/]+)/ (?# group 1 here can be null if group 2 matches)
+		|                         (?# the alternation making it so that only either should match)
+		git@([^:]+):/?            (?# group 2 here can be null if group 1 matches)
+	)
+	(?# removing what follows makes the two first groups nullable, although it then has group 2 unsettable which looks buggy too as PREG_UNMATCHED_AS_NULL is present)
```

Comment:
@Seldaek not sure what this means. could you come up with a concrete code sample and assertions for it?

Reply:
I just meant if you delete the lines of regex below this comment, then the type of the groups 1 and 2 become correct.

Reply:
I see thanks. I will remove the comment from the test so it not confuses others :)